src/topology.jl

```diff
@@ -0,0 +1,170 @@
+@expose_data graph begin
+    property(topology)
+    ref(m -> m.topology)
+    get(m -> deepcopy(m.topology))
+end
+
+# Convenience local aliases.
+const T = Topologies
+const U = Topologies.Unchecked
+const imap = Iterators.map
+const ifilter = Iterators.filter
+
+# Re-export from Topologies.
+export disconnected_components
+
+include("./basic_topology_queries.jl")
+
+"""
+    remove_species!(g::Topology, node::Symbol)
+
+Remove the given species from the topology.
+The species name will remain in-place so that integer-based-indexing remains stable,
+but it will be replaced with a tombstone
+and all incoming and outgoing links will be forgotten.
+"""
+function remove_species!(g::Topology, node::T.IRef)
+    check_species(g)
+    T.remove_node!(g, T.relative(node), :species)
+end
+export remove_species!
+
+"""
+    restrict_to_live_species!(g::Topology, biomasses; threshold = 0)
+
+Remove species nodes until only species with "live" biomasses remain
+*i.e.* biomasses above the given threshold.
+"""
+function restrict_to_live_species!(g::Topology, biomasses; threshold = 0)
+    # Rely on species index memory to map biomass values to labels,
+    # even though some species may already have been removed.
+    check_species(g)
+    sp = U.node_type_index(g, :species)
+    no = U.n_nodes_including_removed(g, sp)
+    nb = length(biomasses)
+    if no != nb
+        n_live = U.n_nodes(g, sp)
+        rem = n_live == no ? "" : " ($(no - n_live) removed)"
+        argerr("The given topology indexes $no species$rem, \
+                but the given biomasses vector size is $nb.")
+    end
+    for (i_sp, bm) in zip(U.nodes_abs_indices(g, sp), biomasses)
+        rm = U.is_removed(g, i_sp)
+        if bm > threshold
+            rm && argerr("Species $(repr(U.node_label(g, i_sp))) \
+                          has been removed from this topology, \
+                          but its biomass is still above threshold: $bm > $threshold.")
+        else
+            rm || T._remove_node!(g, i_sp, sp)
+        end
+    end
+    g
+end
+export restrict_to_live_species!
+
+"""
+    isolated_producers(m::Model, g::Topology)
+
+Iterate over isolated producers nodes in the topology
+*i.e.* producers without incoming or outgoing edges.
+"""
+function isolated_producers(m::InnerParms, g::Topology)
+    sp = U.node_type_index(g, :species)
+    abs(i_rel) = U.node_abs_index(g, T.Rel(i_rel), sp)
+    unwrap(i) = i.abs
+    imap(unwrap, ifilter(imap(abs, get_producers_indices(m))) do i_prod
+        inc = g.incoming[i_prod.abs]
+        inc isa T.Tombstone && return false
+        any(!isempty, inc) && return false
+        out = g.outgoing[i_prod.abs]
+        any(!isempty, out) && return false
+        true
+    end)
+end
+@method isolated_producers depends(Foodweb)
+export isolated_producers
+
+"""
+    starving_consumers(m::Model, g::Topology)
+
+Iterate over starving consumers nodes in the topology
+*i.e.* consumers with no directed trophic path to a consumer.
```

Comment:
*i.e.* consumers with no directed trophic path to a ~consumer~ _producer_.

Reply:
Wops, indeed ^ ^" Fixed. Thank you.